### ISSUE TITLE
ipvs NodePort is not working if host not using eth{i} as nic name

### ISSUE THREAD

**Issue by gyliu513** [kubernetes/kubernetes#55923]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>

/kind bug

> /kind feature

**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

1) Install kubernetes with https://github.com/kubernetes/kubernetes/tree/master/pkg/proxy/ipvs#local-up-cluster

2) After the cluster is ready, check ipvs is working well

```
root@k8s001:~/cases# ipvsadm -ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  10.0.0.1:443 rr persistent 10800
  -> 10.0.3.15:6443               Masq    1      1          1
TCP  10.0.0.10:53 rr
  -> 172.17.0.2:53                Masq    1      0          0
UDP  10.0.0.10:53 rr
  -> 172.17.0.2:53                Masq    1      0          0
```

3) Create a service with NodePort

```
root@k8s001:~/cases# cat  nginx.yaml.2
apiVersion: v1
kind: Service
metadata:
  labels:
    run: my-nginx
  name: my-nginx
  namespace: default
spec:
  ports:
  - name: nginx
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    run: my-nginx
  type: NodePort
---
apiVersion: apps/v1beta1
kind: Deployment
metadata:
  labels:
    run: my-nginx
  name: my-nginx
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      run: my-nginx
  template:
    metadata:
      labels:
        run: my-nginx
    spec:
      containers:
      - image: nginx:1.8.1
        imagePullPolicy: IfNotPresent
        name: my-nginx
        ports:
        - containerPort: 80
          protocol: TCP
```

4) After the service created, check the services.

Pod is running
```
root@k8s001:~/cases# kc get pods
NAME                        READY     STATUS    RESTARTS   AGE
my-nginx-6799bc5758-r75dm   1/1       Running   0          4m
```

service is running
```
root@k8s001:~/cases# kc get svc
NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)        AGE
kubernetes   ClusterIP   10.0.0.1     <none>        443/TCP        8m
my-nginx     NodePort    10.0.0.176   <none>        80:30750/TCP   5m
```

ClusterIP is working well
```
root@k8s001:~/cases# curl 10.0.0.176
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

```
root@k8s001:~/cases# ipvsadm -ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  10.0.0.1:443 rr persistent 10800
  -> 10.0.3.15:6443               Masq    1      1          0
TCP  10.0.0.10:53 rr
  -> 172.17.0.2:53                Masq    1      0          0
TCP  10.0.0.176:80 rr
  -> 172.17.0.3:80                Masq    1      0          1
UDP  10.0.0.10:53 rr
  -> 172.17.0.2:53                Masq    1      0          0
```

NodePort hang  as following:
```
root@k8s001:~/cases# curl 192.168.56.11:30750

```

/sig network

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by stewart-yu** [comment]
I will help to fix it.

**Comment by cmluciano** [comment]
@kubernetes/sig-network-bugs

**Comment by m1093782566** [comment]
@gyliu513 Can you check the eth{i} address on your host?

I guess none of your eth{i} has the IP `192.168.56.11`.

**Comment by gyliu513** [comment]
@m1093782566 I do have the nic for `192.168.56.11` but it was named as `enp0s3`, must we use `eth{i}` for ipvs?

```
root@k8s001:~/cases# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:59:51:1a brd ff:ff:ff:ff:ff:ff
    inet 192.168.56.11/24 brd 192.168.56.255 scope global enp0s3
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe59:511a/64 scope link
       valid_lft forever preferred_lft forever
3: enp0s8: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 08:00:27:0d:b7:1d brd ff:ff:ff:ff:ff:ff
    inet 10.0.3.15/24 brd 10.0.3.255 scope global enp0s8
       valid_lft forever preferred_lft forever
    inet6 fe80::a00:27ff:fe0d:b71d/64 scope link
       valid_lft forever preferred_lft forever
4: enp0s9: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 08:00:27:3b:23:2c brd ff:ff:ff:ff:ff:ff
5: br-1d03525a5f5a: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:f1:08:81:c8 brd ff:ff:ff:ff:ff:ff
    inet 172.20.0.1/16 scope global br-1d03525a5f5a
       valid_lft forever preferred_lft forever
6: br-3584a4451037: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:32:21:9d:b9 brd ff:ff:ff:ff:ff:ff
    inet 172.19.0.1/16 scope global br-3584a4451037
       valid_lft forever preferred_lft forever
7: docker0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:3e:8a:21:f7 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:3eff:fe8a:21f7/64 scope link
       valid_lft forever preferred_lft forever
8: br-9a3a1e266436: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:e9:c6:c3:18 brd ff:ff:ff:ff:ff:ff
    inet 172.21.0.1/16 scope global br-9a3a1e266436
       valid_lft forever preferred_lft forever
9: br-f31235b5511e: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:70:90:e3:9c brd ff:ff:ff:ff:ff:ff
    inet 172.18.0.1/16 scope global br-f31235b5511e
       valid_lft forever preferred_lft forever
11: dummy0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether ee:7b:48:77:56:0c brd ff:ff:ff:ff:ff:ff
12: kube-ipvs0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default
    link/ether 36:3a:e7:8a:e9:da brd ff:ff:ff:ff:ff:ff
    inet 10.0.0.1/32 brd 10.0.0.1 scope global kube-ipvs0
       valid_lft forever preferred_lft forever
    inet 10.0.0.10/32 brd 10.0.0.10 scope global kube-ipvs0
       valid_lft forever preferred_lft forever
    inet 10.0.0.107/32 brd 10.0.0.107 scope global kube-ipvs0
       valid_lft forever preferred_lft forever
14: veth5194ce8@if13: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default
    link/ether 4a:94:c7:73:30:fb brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet6 fe80::4894:c7ff:fe73:30fb/64 scope link
       valid_lft forever preferred_lft forever
16: vethb1ad416@if15: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master docker0 state UP group default
    link/ether 32:9e:08:f6:00:1e brd ff:ff:ff:ff:ff:ff link-netnsid 1
    inet6 fe80::309e:8ff:fef6:1e/64 scope link
       valid_lft forever preferred_lft forever
```

**Comment by m1093782566** [comment]
> must we use eth{i} for ipvs?

Yes. There is a design restriction for now - I know it's challenging and we are trying to get rid of it.

**Comment by gyliu513** [comment]
Thanks @m1093782566 , it is working now after change the nic to `eth{i}`.

```
root@k8s001:~/cases# ipvsadm -ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  192.168.56.11:32647 rr
  -> 172.17.0.3:80                Masq    1      0          0
TCP  10.0.0.1:443 rr persistent 10800
  -> 10.0.3.15:6443               Masq    1      1          0
TCP  10.0.0.10:53 rr
  -> 172.17.0.2:53                Masq    1      0          0
TCP  10.0.0.253:80 rr
  -> 172.17.0.3:80                Masq    1      0          0
TCP  10.0.3.15:32647 rr
  -> 172.17.0.3:80                Masq    1      0          0
UDP  10.0.0.10:53 rr
  -> 172.17.0.2:53                Masq    1      0          0
TCP  [fe80::a00:27ff:fe0d:b71d]:32647 rr
  -> [ac11:3::]:80                Masq    1      0          0
TCP  [fe80::a00:27ff:fe59:511a]:32647 rr
  -> [ac11:3::]:80                Masq    1      0          0
```

```
root@k8s001:~/cases# curl 192.168.56.11:32647
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
```

So does there are any document mentioning this limitation? Otherwise, I'd like to keep this open till we have a fix. Thanks.

**Comment by m1093782566** [comment]
I just come up with an idea about better implementing nodeport type service via ipvs - get rid of eth{i} assumption. Can we take all the IP address which ADDRTYPE matches dst-type LOCAL as the address of ipvs service? For example,

```
[root@100-106-179-225 ~]# ip route show table local type local proto kernel
100.106.179.225 dev eth0  proto kernel  scope host  src 100.106.179.225 
127.0.0.0/8 dev lo  proto kernel  scope host  src 127.0.0.1 
127.0.0.1 dev lo  proto kernel  scope host  src 127.0.0.1 
172.16.0.0 dev flannel.1  proto kernel  scope host  src 172.16.0.0 
172.17.0.1 dev docker0  proto kernel  scope host  src 172.17.0.1 
192.168.122.1 dev virbr0  proto kernel  scope host  src 192.168.122.1 
10.64.77.3 dev kube-ipvs0 proto kernel  scope host  src 192.168.122.1 
```

Because we bind Cluster IPs(e.g `10.64.77.3` in our example) to the dummy interface in IPVS proxier, we need to filter the them out so that we can get the Node IPs. 
Fortunately, the dummy interface created by IPVS proxier is known as kube-ipvs0, so we just need to specify the `dev kube-ipvs0` flag in ip route command, for example,

```
[root@100-106-179-225 ~]# ip route show table local type local proto kernel dev kube-ipvs0
10.64.77.3 dev kube-ipvs0 proto kernel  scope host  src 192.168.122.1 
```

I have already initiated a discussion on sig-network mailing list. Will start to work towards it if community guys are happy with the proposal.

Any comments are welcome!

**Comment by m1093782566** [comment]
After discussing in sig-network, I think I already have a clear map. 

Let's fix it.

/assign

**Comment by m1093782566** [comment]
I believe this can be fixed by #56685 .
